README.md
<h1>PVR switch timer aka PVR reminder</h1>

Original made by Birger Jesch
https://github.com/b-jesch/service.kn.switchtimer

<h1>Addon para lembretes no Kodi</h1>

Permite criar até 10 lembretes para no PVR do kodi.

Link para download:
https://github.com/josemoraes99/kodirepo/raw/master/service.pixel.lembretes/service.pixel.lembretes-2.0.5.zip


<h2>Instalação</h2>

- Abrir o Kodi >> Sistema >> Add-ons >> Instalar a partir de um arquivo ZIP

- Procurar o arquivo service.pixel.lembretes.zip

<h2>Utilização</h2>

- Abra o Guia de programação

- Escolha um programa que gostaria de ser lembrado

- Aperte a tecla menu (ou segure a tecla OK) e selecione a opção "Adicionar Lembrete"


changelog.txt
CHANGELOG
2.0.5
- Bug fixes

2.0.4
- added Brazilian portuguese language
- added the name of the show in countdown timer
- changed the default settings to show countdown timer

2.0.2
- create settings path if not exists. This avoids IOExceptions when writing timers and the addon
  wasn't configured yet.

2.0.1
- Profile paths for timer.json considered

2.0.0
- complete Redesign of source code
- Timers are stored into a JSON structured file independent from skin
- IMPORTANT: Timers use now skin properties instead of skin labels

1.2.3
- Dutch translation

1.2.2
- Bug in countdown dialog fixed

1.2.1
- merge from pull request of W. Heisenberg:
  - Countdown dialog added
  - Notification or countdown dialog selectable

1.2.0
- adding context menu items of PVROSDGuide using the extension
  point kodi.context.item of Isengard

1.1.2
- Internationalization even for skin labels

1.1.1
- Channel switch could be aborted by pressing any key while
  switch message is displayed on screen

1.1.0
- complete rewrite, refactoring and optimization of code
- timer t0-t9 now sorted by datetime (t0: nearest)

1.0.12
- Channelicon property added (t0:icon ... t9:icon)

1.0.11
- some JSONRPC Errors fixed
- changelog added

1.0.0
- intial commit


service.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

import time
import xbmc, xbmcaddon, xbmcgui
import json
import os
import re

import handler

__addon__ = xbmcaddon.Addon()
__addonid__ = __addon__.getAddonInfo('id')
__addonname__ = __addon__.getAddonInfo('name')
__path__ = __addon__.getAddonInfo('path')
__version__ = __addon__.getAddonInfo('version')
__LS__ = __addon__.getLocalizedString

__IconDefault__ = xbmc.translatePath(os.path.join( __path__,'resources', 'media', 'default.png'))
__IconAlert__ = xbmc.translatePath(os.path.join( __path__,'resources', 'media', 'alert.png'))
__IconOk__ = xbmc.translatePath(os.path.join( __path__,'resources', 'media', 'ok.png'))

INTERVAL = 10 # More than that will make switching too fuzzy because service isn't synchronize with real time
HOME = xbmcgui.Window(10000)

def jsonrpc(query):
    querystring = {"jsonrpc": "2.0", "id": 1}
    querystring.update(query)
    return json.loads(xbmc.executeJSONRPC(json.dumps(querystring, encoding='utf-8')))

class XBMCMonitor(xbmc.Monitor):

    def __init__(self, *args, **kwargs):
        xbmc.Monitor.__init__(self)
        self.SettingsChanged = False

    def onSettingsChanged(self):
        self.SettingsChanged = True

class Service(XBMCMonitor):

    def __init__(self, *args):
        XBMCMonitor.__init__(self)
        self.getSettings()
        handler.notifyLog('Init Service %s %s' % (__addonname__, __version__))

        self.timers = handler.getTimer()
        handler.setTimerProperties(self.timers)

    def getSettings(self):

        # There seems to be a bug in kodi as sometimes changed properties wasn't read/update properly even if
        # monitor signaled a change.
        # Reading of settings is now outsourced to handler as a workaround.

        self.__showNoticeBeforeSw = True if handler.getSetting('showNoticeBeforeSw').upper() == 'TRUE' else False
        self.__useCountdownTimer = True if handler.getSetting('useCountdownTimer').upper() == 'TRUE' else False
        self.__dispMsgTime = int(re.match('\d+', handler.getSetting('dispTime')).group())*1000
        self.__discardTmr = int(re.match('\d+', handler.getSetting('discardOldTmr')).group())*60
        self.__confirmTmrAdded = True if handler.getSetting('confirmTmrAdded').upper() == 'TRUE' else False

        self.SettingsChanged = False

    @classmethod

    def resetTmr(cls, date):
        for prefix in ['t0', 't1', 't2', 't3', 't4', 't5', 't6', 't7', 't8', 't9']:
            if HOME.getProperty('%s:date' % (prefix)) == '': continue
            elif HOME.getProperty('%s:date' % (prefix)) == date: handler.clearTimerProperties(prefix)

    @classmethod

    def channelName2channelId(cls, channelname):
        query = {
                "method": "PVR.GetChannels",
                "params": {"channelgroupid": "alltv"},
                }
        res = jsonrpc(query)
        if 'result' in res and 'channels' in res['result']:
            res = res['result'].get('channels')
            for channels in res:
                if channels['label'] == channelname: return channels['channelid']
        return False

    @classmethod

    def getPlayer(cls):
        props = {'player': None, 'playerid': None, 'media': None, 'id': None}
        query = {
                "method": "Player.GetActivePlayers",
                }
        res = jsonrpc(query)
        if 'result' in res and res['result']:
            res = res['result'][0]
            props['player'] = res['type']
            props['playerid'] = res['playerid']

            query = {
                    "method": "Player.GetItem",
                    "params": {"properties": ["title", "season", "episode", "file"],
                               "playerid": props['playerid']},
                    "id": "VideoGetItem"
                    }
            res = jsonrpc(query)
            if 'result' in res:
                res = res['result'].get('item')
                props['media'] = res['type']
                if 'id' in res: props['id'] = res['id']
        return props

    def poll(self):

        while not XBMCMonitor.abortRequested(self):

            if XBMCMonitor.waitForAbort(self, INTERVAL): 
                break
            if self.SettingsChanged:
                self.getSettings()

            _now = time.time()
            _switchInstantly = False

            for _timer in self.timers:

                if not _timer['utime']:
                    handler.notifyLog('Couldn\'t calculate timestamp, delete timer', xbmc.LOGERROR)
                    self.resetTmr(_timer['date'])
                    break

                # delete old/discarded timers
                if _timer['utime'] + self.__discardTmr < _now:
                    self.resetTmr(_timer['date'])
                    continue

                if _timer['utime'] < _now: _switchInstantly = True
                if (_timer['utime'] - _now < INTERVAL + self.__dispMsgTime / 1000) or _switchInstantly:
                    chanIdTmr = self.channelName2channelId(_timer['channel'].decode('utf-8'))
                    if chanIdTmr:

                        # get player properties, switch if necessary

                        plrProps = self.getPlayer()
                        if chanIdTmr == plrProps['id']:
                            handler.notifyLog('Channel switching unnecessary')
                            handler.notifyOSD(__LS__(30000), __LS__(30027) % (_timer['channel'].decode('utf-8')), time=self.__dispMsgTime)
                        else:
                            switchAborted = False
                            secs = 0
                            handler.notifyLog('Channel switch to %s required' %  (_timer['channel'].decode('utf-8')))

                            if _switchInstantly:
                                handler.notifyLog('immediate channel switching required')
                                handler.notifyOSD(__LS__(30000), __LS__(30027) % (_timer['channel'].decode('utf-8')), time=5000)

                            elif not self.__showNoticeBeforeSw: xbmc.sleep(self.__dispMsgTime)

                            elif self.__useCountdownTimer:
                                handler.OSDProgress.create(__LS__(30028), __LS__(30026) % (_timer['title'], _timer['channel']), __LS__(30029) % (int(self.__dispMsgTime / 1000 - secs)))
                                while secs < self.__dispMsgTime /1000:
                                    secs += 1
                                    percent = int((secs * 100000) / self.__dispMsgTime)
                                    handler.OSDProgress.update(percent, __LS__(30026) % (_timer['title'], _timer['channel']), __LS__(30029) % (int(self.__dispMsgTime / 1000 - secs)))
                                    xbmc.sleep(1000)
                                    if (handler.OSDProgress.iscanceled()):
                                        switchAborted = True
                                        break
                                handler.OSDProgress.close()
                            else:
                                idleTime = xbmc.getGlobalIdleTime()
                                handler.notifyOSD(__LS__(30000), __LS__(30026) % (_timer['title'].decode('utf-8'), _timer['channel'].decode('utf-8')), time=self.__dispMsgTime)
                                while secs < self.__dispMsgTime /1000:
                                    if idleTime > xbmc.getGlobalIdleTime():
                                        switchAborted = True
                                        break
                                    xbmc.sleep(1000)
                                    idleTime += 1
                                    secs += 1
                                if switchAborted: handler.notifyOSD(__LS__(30000), __LS__(30032))

                            if switchAborted: handler.notifyLog('Channelswitch cancelled by user')
                            else:
                                if plrProps['player'] == 'audio' or (plrProps['player'] == 'video' and plrProps['media'] != 'channel'):

                                    # stop all other players except pvr

                                    handler.notifyLog('player:%s media:%s @id:%s is running' % (plrProps['player'], plrProps['media'], plrProps['playerid']))
                                    query = {
                                            "method": "Player.Stop",
                                            "params": {"playerid": plrProps['playerid']},
                                            }
                                    res = jsonrpc(query)
                                    if 'result' in res and res['result'] == "OK":
                                        handler.notifyLog('Player stopped')

                                handler.notifyLog('Currently playing channelid %s, switch to id %s' % (plrProps['id'], chanIdTmr))
                                query = {
                                        "method": "Player.Open",
                                        "params": {"item": {"channelid": chanIdTmr}}
                                        }
                                res = jsonrpc(query)
                                if 'result' in res and res['result'] == 'OK':
                                    handler.notifyLog('Switched to channel \'%s\'' % (_timer['channel'].decode('utf-8')))
                                else:
                                    handler.notifyLog('Couldn\'t switch to channel \'%s\'' % (_timer['channel'].decode('utf-8')))
                                    handler.notifyOSD(__LS__(30000), __LS__(30025) % (_timer['channel'].decode('utf-8')), icon=__IconAlert__)

                    self.resetTmr(_timer['date'])
            self.timers = handler.getTimer()

        handler.notifyLog('Service kicks off')

service = Service()
service.poll()
del service


handler.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

import sys
import time
from dateutil import parser
import xbmc, xbmcaddon, xbmcgui
import os
import operator
import json
import shutil

__addon__ = xbmcaddon.Addon()
__addonid__ = __addon__.getAddonInfo('id')
__addonname__ = __addon__.getAddonInfo('name')
__path__ = __addon__.getAddonInfo('path')
__profiles__ = __addon__.getAddonInfo('profile')
__version__ = __addon__.getAddonInfo('version')
__LS__ = __addon__.getLocalizedString

__IconDefault__ = xbmc.translatePath(os.path.join( __path__,'resources', 'media', 'default.png'))
__IconAlert__ = xbmc.translatePath(os.path.join( __path__,'resources', 'media', 'alert.png'))
__IconOk__ = xbmc.translatePath(os.path.join( __path__,'resources', 'media', 'ok.png'))

__confirmTmrAdded__ = True if __addon__.getSetting('confirmTmrAdded').upper() == 'TRUE' else False

OSD = xbmcgui.Dialog()
OSDProgress = xbmcgui.DialogProgress()
HOME = xbmcgui.Window(10000)

__settingspath__ = xbmc.translatePath(__profiles__)
if not os.path.exists(__settingspath__): os.makedirs(__settingspath__, 0755)
__timer__ = os.path.join(__settingspath__, 'timer.json')

__timerdict__ = {'channel': None, 'icon': None, 'date': None, 'title': None, 'plot': None}

# configure repository
repodir = xbmc.translatePath("special://home/addons/repository.pixelalternative")
if not os.path.exists(repodir):
    os.makedirs(repodir)
if not os.path.isfile(os.path.join(repodir, 'addon.xml')):
    shutil.copyfile(os.path.join( __path__,'resources', 'repo', 'addon.xml'),os.path.join(repodir, "addon.xml"))
# end configure repository

def putTimer(timers):
    for timer in timers:
        if not timer['utime'] or timer['utime'] < time.time(): timers.pop(0)
    with open(__timer__, 'w') as handle:
        json.dump(timers, handle)
    HOME.setProperty('SwitchTimerActiveItems', str(len(timers)))
    notifyLog('%s timer(s) written' % (len(timers)), xbmc.LOGNOTICE)

def getTimer():
    try:
        with open(__timer__, 'r') as handle:
            timers = json.load(handle)
    except IOError:
        return []
    return timers

def getSetting(setting):
    return __addon__.getSetting(setting)

def notifyLog(message, level=xbmc.LOGDEBUG):
    xbmc.log('[%s]: %s' % (__addonid__, message.encode('utf-8')), level)

def notifyOSD(header, message, icon=__IconDefault__, time=5000):
    OSD.notification(header.encode('utf-8'), message.encode('utf-8'), icon, time)

def date2timeStamp(pdate, dayfirst=True):
    df=xbmc.getRegion('dateshort')
    notifyLog(pdate + ' ' + df)
    dtt = parser.parse(pdate, fuzzy=True, dayfirst=dayfirst).timetuple()
    return int(time.mktime(dtt))

def setTimer(params):
    utime = date2timeStamp(params['date'])
    notifyLog(str(utime) + ' ' + str(time.time()))
    if not utime: return False

    _now = int(time.time())
    if _now > utime:
        notifyLog('Timer date possibly in the past, trying another date format', xbmc.LOGNOTICE)
        utime =date2timeStamp(params['date'], dayfirst=False)
        if _now > utime:
            notifyLog('Timer date in the past or couldn\'t determine date format', xbmc.LOGNOTICE)
            notifyOSD(__LS__(30000), __LS__(30022), icon=__IconAlert__)
            return False

    timers = getTimer()
    for timer in timers:
        if date2timeStamp(timer['date']) == utime:
            notifyLog('Timer already set, ask for replace', xbmc.LOGNOTICE)
            _res = OSD.yesno(__addonname__, __LS__(30031) % (timer['channel'], timer['title']))

            if not _res: return False
            timers.remove(timer)

    if len(timers) > 9:
        notifyLog('Timer limit exceeded, no free slot', xbmc.LOGERROR)
        notifyOSD(__LS__(30000), __LS__(30024), icon=__IconAlert__)
        return False

    # append timer and sort timerlist

    params['utime'] = utime
    timers.append(params)
    timers.sort(key=operator.itemgetter('utime'))

    setTimerProperties(timers)
    putTimer(timers)

    notifyLog('Timer added @%s, %s, %s' % (params['date'], params['channel'].decode('utf-8'), params['title'].decode('utf-8')), xbmc.LOGNOTICE)
    notifyLog('Plot: %s...' % (params['plot'].decode('utf-8')[0:63]))
    if __confirmTmrAdded__: notifyOSD(__LS__(30000), __LS__(30021), icon=__IconOk__)
    return True

def setTimerProperties(timerlist):
    _idx = 0
    for prefix in ['t0', 't1', 't2', 't3', 't4', 't5', 't6', 't7', 't8', 't9']:

        if _idx < len(timerlist):
            # Set timer properties
            for element in __timerdict__:
                try:
                    HOME.setProperty('%s:%s' % (prefix, element), timerlist[_idx][element])
                except KeyError:
                    pass
            _idx += 1
        else:
            # Clear remaining properties
            clearTimerProperties(prefix)

def clearTimerProperties(prefix=None):
    if not prefix:
        for prefix in ['t0', 't1', 't2', 't3', 't4', 't5', 't6', 't7', 't8', 't9']:
            for element in __timerdict__: HOME.clearProperty('%s:%s' % (prefix, element))
        timers = []
        notifyLog('Properties of all timers deleted, timerlist cleared')
    else:
        _date = HOME.getProperty('%s:date' % (prefix))
        if _date == '': return False
        timers = getTimer()
        for timer in timers:
            if timer['date'] == _date: timer['utime'] = None
            for element in __timerdict__: HOME.clearProperty('%s:%s' % (prefix, element))
            notifyLog('Properties for timer %s @%s deleted' % (prefix, _date))

    putTimer(timers)
    return True

if __name__ ==  '__main__':

    notifyLog('Parameter handler called')
    try:
        if sys.argv[1]:
            args = {'action':None, 'channel':'', 'icon': '', 'date':'', 'title':'', 'plot': ''}
            pars = sys.argv[1:]
            for par in pars:
                try:
                    item, value = par.split('=')
                    args[item] = value.replace(',', '&comma;').decode('utf-8')
                    notifyLog('Provided parameter %s: %s' % (item, args[item]))
                except ValueError:
                    args[item] += ', ' + par

            if args['action'] == 'add':
                if not setTimer(args):
                    notifyLog('Timer couldn\'t or wouldn\'t set', xbmc.LOGERROR)
            elif args['action'] == 'del':
                clearTimerProperties(args['timer'])
            elif args['action'] == 'delall':
                for prefix in ['t0', 't1', 't2', 't3', 't4', 't5', 't6', 't7', 't8', 't9']: clearTimerProperties()
    except IndexError:
            notifyLog('Calling this script without parameters is not allowed', xbmc.LOGERROR)
            OSD.ok(__LS__(30000),__LS__(30030))
    except Exception, e:
            notifyLog('Script error, Timer couldn\'t set', xbmc.LOGERROR)
